### ISSUE TITLE
gasnet@2022.3.0 +cuda build fails: cc1: error: invalid --param name 'max-inline-insns'

### ISSUE THREAD

**Issue by eugeneswalker** [LLNL/spack#30589]
### Steps to reproduce the issue

`gasnet@2022.3.0 +cuda` build fails using:
* `spack/develop` (a65e00392c0cd6a573937afc01b7b80d075946f1 from `Tue May 10 09:05:19 2022 -0600`)
* Ubuntu 20.04, X86_64
* GCC 9.4.0
* CUDA 11.4.2

Concrete spec: [gasnet-mifp3w5.spec.json.txt](https://github.com/spack/spack/files/8662945/gasnet-mifp3w5.spec.json.txt)

Concretization:
```
 -   gasnet@2022.3.0%gcc@9.4.0+cuda~debug~rocm conduits=smp cuda_arch=80 arch=linux-ubuntu20.04-x86_64
[+]      ^cuda@11.4.2%gcc@9.4.0~allow-unsupported-compilers~dev arch=linux-ubuntu20.04-x86_64
[+]          ^libxml2@2.9.13%gcc@9.4.0~python arch=linux-ubuntu20.04-x86_64
[+]              ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]              ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]              ^xz@5.2.5%gcc@9.4.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]              ^zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
```

Quickly reproducible using cached dependency binaries and Docker container image:
* `ecpe4s/ubuntu20.04-runner-x86_64:2022-05-01`

```
$> git clone https://github.com/spack/spack
$> (cd spack && git checkout a65e0039)
$> . spack/share/spack/setup-env.sh

$> spack mirror add E4S https://cache.e4s.io
$> spack buildcache keys -it

$> spack install --cache-only --include-build-deps --only dependencies -f ./gasnet-mifp3w5.spec.json

$> spack install --no-cache -f ./gasnet-mifp3w5.spec.json
...
# SEE ERROR MESSAGE BELOW
```

### Error message

<details><summary>Error message</summary><pre>
==> Installing gasnet-2022.3.0-mifp3w53mba3nmhkzsyke53zsapophir
==> Fetching https://mirror.spack.io/_source-cache/archive/91/91b59aa84c0680c807e00d3d1d8fa7c33c1aed50b86d1616f93e499620a9ba09.tar.gz
==> No patches needed for gasnet
==> gasnet: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    './configure' '--prefix=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/gasnet-2022.3.0-mifp3w53mba3nmhkzsyke53zsapophir' '--enable-kind-cuda-uva' '--disable-mpi-compat' '--disable-auto-conduit-detect' '--enable-smp' '--enable-rpath'

2 errors found in build log:
     85     checking for strings.h... yes
     86     checking for inttypes.h... yes
     87     checking for stdint.h... yes
     88     checking for unistd.h... yes
     89     checking size of void *... 8
     90     checking for C compiler flag --param max-inline-insns-single=35000... yes
  >> 91     checking for C compiler flag --param max-inline-insns=100000... no/error: cc1: error: invalid --param name 'max-inline-insns'; did
            you mean 'max-combine-insns'?
     92     checking for C compiler flag --param inline-unit-growth=10000... yes
     93     checking for C compiler flag --param large-function-growth=200000... yes
     94     checking for C compiler flag -Winline... yes
     95     checking for buggy -Winline... ok
     96     checking for working C99 printf specifiers... yes
     97     checking for C compiler flag -Wall... yes

     ...

     430    checking for CUDA_LIBS setting...          (default)  "-lcuda"
     431    checking for CUDA_LDFLAGS setting...       (default)  ""
     432    checking for working CUDA configuration... no
     433
     434    configure error: CUDA UVA memory kinds support was requested with --enable-kind-cuda-uva but is not available
     435
  >> 436    configure: error: See /tmp/root/spack-stage/spack-stage-gasnet-2022.3.0-mifp3w53mba3nmhkzsyke53zsapophir/spack-src/config.log for d
            etails.

See build log for details:
  /tmp/root/spack-stage/spack-stage-gasnet-2022.3.0-mifp3w53mba3nmhkzsyke53zsapophir/spack-build-out.txt
</pre></details>

### Information on your system

* **Spack:** 0.18.0.dev0 (a65e00392c0cd6a573937afc01b7b80d075946f1)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/8663017/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/8663016/spack-build-env.txt)

@PHHargrove @bonachea @wspear 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by bonachea** [comment]
Hi @eugeneswalker, here is the actual error message from the config.log you did not include:

```
configure:262535: /spack/lib/spack/env/gcc/gcc -o conftest -O3 --param max-inline-insns-single=35000 --param inline-unit-growth=10000 --param large-function-growth=200000  -Wno-unused -Wunused-result -Wno-unused-parameter -Wno-address -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/cuda-11.4.2-2crhncdchytyasbmyj6bvs7nylbep6fr/include -D_GNU_SOURCE=1    conftest.c  -lcuda >&5
/usr/bin/ld: cannot find -lcuda
collect2: error: ld returned 1 exit status
```

The problem is GASNet's `+cuda` variant requires the CUDA *Driver* Library, the hardware-specific CUDA driver which is not included in the CUDA Runtime Library distribution (Spack's CUDA package). See the second sentence of the description for the CUDA spackage:

```
root@e9bb2d9ceb6c:/# spack info cuda
Package:   cuda

Description:
    CUDA is a parallel computing platform and programming model invented by
    NVIDIA. It enables dramatic increases in computing performance by
    harnessing the power of the graphics processing unit (GPU). Note: This
    package does not currently install the drivers necessary to run CUDA.
    These will need to be installed manually. See:
    https://docs.nvidia.com/cuda/ for details.

```

I'm not sure how other CUDA-enabled spackages handle this, but most of those probably rely on the CUDA Runtime library.
However GASNet is a networking library that interfaces with low-level GPU functionality like GPUDirect, so `gasnet+cuda` has a hard requirement that the CUDA Driver library be present on the system at build+run time. As such you'll need to follow the instructions above and ensure the CUDA drivers are manually installed for any system where you want to use `gasnet+cuda` (incidentally the same applies to `upcxx+cuda` for the same reasons). 

It's worth noting that for the particular spec you chose: `gasnet +cuda conduits=smp` the `+cuda` dependency is entirely superfluous, because GASNet's smp-conduit does not *currently* feature CUDA integration. So the variant's demand for CUDA support is causing `configure` to fail, but even if `configure` had passed it would not affect the library build because smp-conduit currently ignores CUDA. 

**TL;DR: For this use case I suggest you remove the `+cuda` variant because it's (currently) meaningless for smp-conduit and just introduces a pointless dependency on manually installed CUDA drivers.**

**Comment by eugeneswalker** [comment]
Thank you for that context. Sorry for missing that. Would `conduits=mpi` make more sense to pair with `+cuda` ?

**Comment by bonachea** [comment]
> Thank you for that context. Sorry for missing that. Would `conduits=mpi` make more sense to pair with `+cuda` ?

No, currently the only GASNet conduits that integrate with CUDA GPUDirect are ibv-conduit and ucx-conduit.

The UPC++ package supports `+cuda` for any backend, but also still requires the CUDA Driver libraries be present on the build system (usually part of the distro or manually installed by an admin).

PR #30602 adds a note about this issue to the variant descriptions, to help prevent similar confusion in the future.
